### PR TITLE
Jenkinsfile to replace js_pr

### PR DESCRIPTION
**Pull request by alebsack**
This should be a starting point for replacing `js_pr` in CI.  It's not aimed at fixing all of the failing scenarios, but I'd like to get this in as a way to move forward with fixing them.

@emanuelez @alazier @radu-tutueanu @nhachicha

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
 
-RUN apt-get update && \
-    apt-get install -y curl && \
+# Install the JDK
+# We are going to need some 32 bit binaries because aapt requires it
+# file is need by the script that creates NDK toolchains
+ENV DEBIAN_FRONTEND noninteractive
+RUN dpkg --add-architecture i386 && \
+    apt-get update -qq && \
+    apt-get install -y file git curl wget zip unzip bsdmainutils strace lsof \
```
Comment (emanuelez) [inline Dockerfile]: I prefer to have one package per line and in alphabetical order. That usually helps while checking if a particular package is installed and avoiding duplicates.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      } else {
+        echo "Building release: '${gitTag}'"
+        setBuildName("Tag ${gitTag}")
+      }
+    }
+    echo "version: ${version}"
+  }
+}
+
+def reportStatus(target, state, message) {
```
Comment (emanuelez) [inline Jenkinsfile]: Why do we need this?

Comment (alebsack) [inline Jenkinsfile]: We can set the individual status of each branch of execution in the PR.

File: Dockerfile [hunk 3]
```diff
@@ -1,8 +1,34 @@
 FROM ubuntu:xenial
 
-RUN apt-get update && \
-    apt-get install -y curl && \
+# Install the JDK
+# We are going to need some 32 bit binaries because aapt requires it
+# file is need by the script that creates NDK toolchains
+ENV DEBIAN_FRONTEND noninteractive
+RUN dpkg --add-architecture i386 && \
```
Comment (nhachicha) [inline Dockerfile]: We also need to install Android NDK & SDK similar to https://github.com/realm/realm-java/blob/master/Dockerfile for Android

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,142 @@
+#!groovy
```
Comment (nhachicha) [inline Jenkinsfile]: for Android we need to have access to the device to run tests inside, currently this is how `realm-java` is setup https://github.com/realm/realm-java/blob/master/Jenkinsfile#L43

The current android test also extract a file `tests/react-test-app/tests.xml` that the current conf in CI uses to publish Unit Test result
screen shot 2016-11-23 at 18 44 09

File: Dockerfile [hunk 5]
```diff
@@ -1,8 +1,75 @@
 FROM ubuntu:xenial
 
-RUN apt-get update && \
-    apt-get install -y curl && \
+# Install the JDK
+# We are going to need some 32 bit binaries because apt requires it
```
Comment (emanuelez) [inline Dockerfile]: It's actually `aapt`, not `apt`. It stands for Android Asset Packaging Tool.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    $class: 'GitHubCommitStatusSetter',
+    contextSource: [$class: 'ManuallyEnteredCommitContextSource', context: target],
+    statusResultSource: [$class: 'ConditionalStatusResultSource', results: [[
+      $class: 'AnyBuildResult', message: message, state: state]]
+    ],
+    reposSource: [$class: 'ManuallyEnteredRepositorySource', url: 'https://github.com/realm/realm-js']
+  ])
+}
+
+def doInside(script, target, postStep = null) {
```
Comment (emanuelez) [inline Jenkinsfile]: Why are we doing docker stuff in bash? Is it something we're gonna need on a developer's desk?

File: Dockerfile (hunk 5 above)
Comment (alebsack) [inline Dockerfile]: That's not confusing at all!

File: Jenkinsfile (hunk 6 above)
Comment (alebsack) [inline Jenkinsfile]: IMO it's functionally the same, but yes, it allows developers to run any of the `test.sh` stuff in a local docker container.

File: Jenkinsfile (hunk 2 above)
Comment (radu-tutueanu) [inline Jenkinsfile]: Nice. Though i guess that if we still see issues with the GitHub api rate limit, we might want to disable it

Comment (alebsack) [inline Jenkinsfile]: I don't think it's making any more GitHub API requests than the existing js_pr job.  If so, we could eliminate the PENDING status.

Comment (radu-tutueanu) [inline Jenkinsfile]: ok, we will see. I don't think js_pr jobs have been an issue until now.

**Comment by alebsack** [comment]
This is RTM.  The MacOS tests are still a nondeterministic.  🎱

**Pull request merged**